kv/tikv/storage/server.go

```diff
@@ -0,0 +1,178 @@
+package storage
+
+import (
+	"context"
+	"github.com/pingcap-incubator/tinykv/kv/pd"
+	"github.com/pingcap-incubator/tinykv/kv/tikv/dbreader"
+	"github.com/pingcap-incubator/tinykv/kv/tikv/inner_server"
+	"github.com/pingcap-incubator/tinykv/kv/tikv/storage/kvstore"
+	"github.com/pingcap-incubator/tinykv/kv/tikv/storage/latches"
+	"github.com/pingcap-incubator/tinykv/proto/pkg/coprocessor"
+	"github.com/pingcap-incubator/tinykv/proto/pkg/kvrpcpb"
+	"github.com/pingcap-incubator/tinykv/proto/pkg/tikvpb"
+)
+
+// InnerServer represents the internal-facing server part of TinyKV, it handles sending and receiving from other
+// TinyKV nodes. As part of that responsibility, it also reads and writes data to disk (or semi-permanent memory).
+type InnerServer interface {
+	Start(pdClient pd.Client) error
+	Stop() error
+	Write(ctx *kvrpcpb.Context, batch []inner_server.Modify) error
+	Reader(ctx *kvrpcpb.Context) (dbreader.DBReader, error)
+	Raft(stream tikvpb.Tikv_RaftServer) error
+	Snapshot(stream tikvpb.Tikv_SnapshotServer) error
+}
+
+var _ tikvpb.TikvServer = new(Server)
+
+// Server is a TinyKV server, it 'faces outwards', sending and receiving messages from clients such as TinySQL.
+type Server struct {
+	innerServer InnerServer
+	Latches     *latches.Latches
+}
+
+func NewServer(innerServer InnerServer) *Server {
+	return &Server{
+		innerServer: innerServer,
+		Latches:     latches.NewLatches(),
+	}
+}
+
+func (server *Server) Run(cmd Command) (interface{}, error) {
+	ctxt := cmd.Context()
+	var resp interface{}
+
+	latches := cmd.WillWrite()
+	if latches == nil {
+		// The command is readonly or requires access to the DB to determine the keys it will write.
+		reader, err := server.innerServer.Reader(ctxt)
+		if err != nil {
+			return nil, err
+		}
+		txn := kvstore.RoTxn{Reader: reader}
+		resp, latches, err = cmd.Read(&txn)
+		reader.Close()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if latches != nil {
+		// The command will write to the DB.
+		for {
+			wg := server.Latches.AcquireLatches(latches)
```

Comment:
Add a function `WaitLatches` for `Latches`?

Reply:
Shall do